### ISSUE TITLE
Deterministic results across platforms

### ISSUE THREAD

**Issue by xoofx** [shibatch/sleef#187]
Hey,
As we discussed about this in the past, I'm just creating this issue to track this down separately.

There are cases where you have to run calculations on different CPU arch (x86 vs ARM) but you still want to achieve exact same binary results for any operations involved (either primary - add/div/sub/mul or more complex like cos/sin...etc.)

This is currently not possible with sleef, but it would be amazing if it could provide such a mode.

Note that this is not about being 100% IEEE 754 compliant, but more to be consistent across popular CPU arch. 

Note also that in order to be relevant and xplat, sleef would have to expose some functions that are not exposed today (like converting -vector-floats to integers, as typically FTZ conventions are not always 100% compatible between CPU arch)

sleef could also support deterministic mode with lower accuracy, as long as it is consistent across platforms.

**Comment by shibatch** [comment]
First of all, making results consistent between aarch32 and other architectures is almost impossible. Making them close to each other could be possible, though.
We definitely need IEEE 754 compliance, and we can only use operators that are available at all platforms. Thus, we cannot use FMA, and that means that 1ULP-accuracy versions of the functions are slow.
In order to make the results of the simulation consistent across all platforms, all the source codes of the application have to be compiled using -ffp-contract=off option. You cannot use -ffast-math.

If all of the above works for you, I can start working on it. Since we need a new design of the API, we have to talk to @fpetrogalli-arm before we start.

**Comment by shibatch** [comment]
@xoofx Any progress?

**Comment by xoofx** [comment]
> If all of the above works for you, I can start working on it. Since we need a new design of the API, we have to talk to @fpetrogalli-arm before we start.
> @xoofx Any progress?

Oops, sorry, yes I'm fine with the requirements.

**Comment by shibatch** [comment]
I am ready to start, but @fpetrogalli-arm seems very busy now.
I think I will need to ask him to review the code many times for this.

@fpetrogalli-arm How do you think?

**Comment by shibatch** [comment]
One big point to consider is whether we assume IEEE 754 compliance.
* If we assume IEEE 754 compliance, it is impossible to get the exactly same results on AArch32. It is hard to estimate how much difference there is between AArch32 and other IEEE 754 compliant computers.
* If we throw away IEEE 754 compliance and enable FTZ on all platforms, consistence is still not perfect.

Another big point is how we will handle SIMD division on AArch32.
There is no division in AArch32 NEON. In SLEEF, it utilizes an approximate division subroutine internally, but it's still approximation. In order to maintain consistency between AArch32 and other platforms, we need a more accurate division subroutine, but this would be slow.

The easiest way is to give up consistency on AArch32.
@xoofx Do you need consistency between AArch32 and other platforms?

**Comment by shibatch** [comment]
For the naming of the functions, I am considering adding various attributes before the current function names. For example,

Sleef_sind4_u10avx512f is the current name.
Sleef_finite_sind4_u10avx512f if we will implement finite-math version of the function.
Sleef_finite_mask_sind4_u10avx512f if we will implement masked and finite-math version of the function. The attributes are arranged in the dictionary order.

And, we could think of global consistency and platform consistency. In this issue, we are basically talking about global consistency. However, in https://github.com/shibatch/sleef/issues/47, we were also talking about consistency, and that was limited within the same platform. So, we could name them "globalconsistent" and "platformconsistent" attributes. Are these too long?

So, the function name of the above function will be Sleef_globalconsistent_sind4_u10avx512.

**Comment by xoofx** [comment]
> The easiest way is to give up consistency on AArch32.
> @xoofx Do you need consistency between AArch32 and other platforms?

AArch32 is still pretty popular but if it is bringing too much trouble for a first iteration, let's focus only on x86/x64/AARch64. With the feedback of this first iteration, we will be able to better reevaluate this. For AArch32, if it requires slow path to get full consistency, that will be acceptable.

> And, we could think of global consistency and platform consistency. In this issue, we are basically talking about global consistency. However, in #47, we were also talking about consistency, and that was limited within the same platform. So, we could name them "globalconsistent" and "platformconsistent" attributes. Are these too long?
> So, the function name of the above function will be Sleef_globalconsistent_sind4_u10avx512.

For the naming, as I'm not calling them directly but this is auto-generated, I don't mind whatever is used.

Concerning determinism, as long as we are able to have u10/u35 versions that's fine (and I would love to see also other options like u60, as we discussed a bit at #107 but that's a separate issue)

**Comment by shibatch** [comment]
u60 won't be much faster than u35.
The difference between u10 and u35 is if it uses DD arithmetic.
Possibly I could implement faster subroutine if it is u1000 or so.
It would be a large help if you could collect statistics as to the frequency of each function to be called.

**Comment by xoofx** [comment]
> u60 won't be much faster than u35.

Oups, indeed, I was actually thinking in terms of bits when writing u60, so yes, it is about u1000. For the functions, it is most basic ones being any cos/sin/tan functions. Do you want me to open a clear separate issue for that?

**Comment by shibatch** [comment]
Do you really need tan?
I thought that atan and acos would be used more frequently.
I also want to know what accuracy you really need.

**Comment by xoofx** [comment]
> Do you really need tan?
> I thought that atan and acos would be used more frequently.

tan is used, but probably not a lot. After checking with a few folks the most important one are probably:

`cos`, `sin`, `log2`/`exp2` (`log`/`exp` but a bit less critical),  `pow`, `acos`, `asin` (from higher to lower priority)

**Comment by xoofx** [comment]
> I also want to know what accuracy you really need.

u10 (23 bits accuracy), u35 (21 bits), u635 (17 bits), u16383 (9 bits), u524287 (4 bits)

**Comment by shibatch** [comment]
And they are all single precision functions?

**Comment by xoofx** [comment]
> And they are all single precision functions?

Yes

**Comment by shibatch** [comment]
How about the input domain of the trig functions?
Do you need wider range than [-2PI, 2PI]?

**Comment by xoofx** [comment]
> How about the input domain of the trig functions?
> Do you need wider range than [-2PI, 2PI]?

Yes, but we could have also an optimized version of [-2PI, 2PI] if it can help

**Comment by fpetrogalli-arm** [comment]
>   For the naming of the functions, I am considering adding various attributes before the current function names. For example,
>
> Sleef_sind4_u10avx512f is the current name.
> Sleef_finite_sind4_u10avx512f if we will implement finite-math version of the function.
> Sleef_finite_mask_sind4_u10avx512f if we will implement masked and finite-math version of the function. The attributes are arranged in the dictionary order.

The current names in libsleef are already too long, making them even longer is even worse. I don't think you can expect people to write those things in user code.

> And, we could think of global consistency and platform consistency. In this issue, we are basically talking about global consistency. However, in #47, we were also talking about consistency, and that was limited within the same platform. So, we could name them "globalconsistent" and "platformconsistent" attributes. Are these too long?
 
Yes :)

This is a very interesting feature and I believe we have to make it as standard as possible, to avoid pushing out people who doin't want to use non standard names (we already know of a couple of important projects that don't want to use custom names). Most of the calls to these vector routines will be auto-generated by a compiler (if not all of them), so I think we should integrate our solution with whatever standard is already available.

In particular, I think we should focus on integrating these new functions in `libsleefgnuabi`.

This should be done in few steps:

1. make sure that all the C names (before renaming) in `sleef[simd][sd]p.c` follow a well defined grammar:

```
<x function name> = x<C99 name>{_<attributes>}_u<ULP>
``` 

2. decide which ULP versiona getselected as the simbol that provides the C99 name in the library - i.e. whether `<C99 name>` is rendered by `<C99 name>_u10` or `<C99 name>_u35`. The local standardization in 1. will make it possible to change the default ULP version even at runtime with a dispatcher.

3. once 1. and 2. are in place, define which `<attributes>` we care about, each attribute being a pair. A  long attribute to be used in the `<x function name>` name in the C source file, and an attribute to single character mapping to be used when rendering the name in the vector function ABI mangled names.

I believe this is all extra work that we should consider if we want to make sure to reach the widest audience of users for SLEEF.

**Comment by shibatch** [comment]
> The current names in libsleef are already too long, making them even longer is even worse. I don't think you can expect people to write those things in user code.

The corresponding function in SVML is _mm512_mask_sin_pd, and this has just mask attribute. 
The important thing is to name them in a consistent, extendable and easy-to-understand way. I think there is only aesthetic problem if the names are too long. People would rename with #define if they are too long.

>This is a very interesting feature and I believe we have to make it as standard as possible, to avoid pushing out people who doin't want to use non standard names (we already know of a couple of important projects that don't want to use custom names). Most of the calls to these vector routines will be auto-generated by a compiler (if not all of them), so I think we should integrate our solution with whatever standard is already available.

The problem is that the vectorizer in today's compilers are not perfect, and I see there is still demand for directly calling SLEEF functions.

>This should be done in few steps:

This looks good. But how do you spell the attributes? I think this plan has the same problem.

**Comment by fpetrogalli-arm** [comment]
>  I think there is only aesthetic problem if the names are too long. People would rename with #define if they are too long.

It is not just an aesthetic problem. Using target specific names reduces the portability of the code, which sometimes is a limiting factor.

> The problem is that the vectorizer in today's compilers are not perfect, 

They will probably never be perfect. But the current vectorizers are ramping up on function vectorization via `#pragma omp declare simd`, it won't take long before most of the uses of the vector math functions will be achieved by writing scalar code.

> and I see there is still demand for directly calling SLEEF functions.

Fair enough.

> But how do you spell the attributes? I think this plan has the same problem.

What do you mean here? Which specific character to use for each attribute? It doesn't really matter, as long as it is properly documented.

**Comment by shibatch** [comment]
>What do you mean here? Which specific character to use for each attribute? It doesn't really matter, as long as it is properly documented.

Do you mean that each attribute is mangled?
Then, the same scheme can be used for the function names beginning with "Sleef_".
My concern is that you need to establish a consensus for the way of naming of GNUABI functions.

**Comment by fpetrogalli-arm** [comment]
> Do you mean that each attribute is mangled?

Yes.

> Then, the same scheme can be used for the function names beginning with "Sleef_".

Yes.

> My concern is that you need to establish a consensus for the way of naming of GNUABI functions.

There is already consensus, the mangling function of the Vector ABI. We will end up having names like `_ZGV[whatever]_<C99 name><attributes>`, and possibly come up with a run-time system that select the one to remap to the C99 standard mangled name  `_ZGV[whatever]_<C99 name>`

**Comment by shibatch** [comment]
@fpetrogalli-arm Who decide the correspondence between a mangled attribute name and an attribute?
Is the error bound (ULP) included in the mangled attributes?

**Comment by shibatch** [comment]
But if attributes are mangled in a different way in each library, there would be no compatibility.

**Comment by shibatch** [comment]
> the user should be able to specify at run time which one to select.

Who is a user in this context? There are developers of many linked libraries, and each developer wants to choose different mapping.
